### ISSUE TITLE
[AUDIT] DEFECT-004: Upstream Core MBSE pipeline lacks automated Ground-Truth Provenance & Anti-Hallucination Scanner

### ISSUE THREAD

**Issue by gintatkinson** [gintatkinson/DEAP01-spec-core#46]
## 1. Context and References

- **File**: `scripts/verify_downstream_baseline.py:750-760`, `skills/spec-orchestrator/`
- **Pillar**: Semantic Traceability
- **Symptom**: Upstream Core MBSE pipeline lacks an automated Ground-Truth Provenance & Anti-Hallucination Scanner in upstream baseline verification (`scripts/verify_downstream_baseline.py`) and specification orchestrator quality gates (`skills/spec-orchestrator/`), allowing ungrounded platform parameter drift (wingspan, MTOW, stall speed, cruise speed, launch pressure), fictitious deep-learning buzzwords (CUDA, YOLO, PyTorch, TensorRT), and broken or unresolved schema citations to propagate undetected into downstream engineering baselines.

## 2. Root Cause Analysis (5 Whys)

1. **Why does the upstream Core MBSE pipeline permit ungrounded specifications and hallucinated architectures?** Because `scripts/verify_downstream_baseline.py` and `skills/spec-orchestrator/` lack an automated quality gate that cross-checks downstream specifications against authoritative engineering ground truth and schema invariants.
2. **Why do baseline verification and specification quality gates omit ground-truth scanning?** Because the existing quality gates (Checks 10–17) only validate file structure, KaTeX math syntax, clean landing zones, and SORA/STPA matrix presence without semantically auditing numerical parameters or detecting ungrounded software/hardware buzzwords.
3. **Why are hallucinated parameters and buzzwords hazardous in downstream engineering?** Because downstream subagents hallucinate contradictory physical parameters (such as 2.4 m wingspan vs 1.8 m SSOT, 25 kg MTOW vs 17.0 kg SSOT) and deep-learning buzzwords (CUDA, PyTorch, YOLOv8) that cannot execute on safety-certified embedded microcontrollers, violating DO-178C Level B and ASTM F3269-17 RTA certifiability.
4. **Why did specification generation subagents introduce ungrounded parameters and buzzwords?** Because LLM subagents exhibit stochastic drift when translating natural language requirements without automated provenance pinning and citation verification against authoritative source documents in `schema/` and `schema/extracted/`.
5. **Why was automated ground-truth provenance verification absent from the upstream core architecture?** Because the upstream distribution template (`DEAP-spec-core`) decoupled abstract MBSE tooling from domain models without providing a platform-invariant ground-truth provenance scanning tool (`scripts/audit_hallucinations.py`) and automated Parity Auditor validator in the core pipeline framework.

## 3. Correctness Analysis

The Digital Engineering Agent Platform (DEAP) relies on formal SysML v2 architectural models (`schema/Avenger5.sysml`) and authoritative technical manuals (`schema/extracted/`) to serve as the Single Source of Truth (SSOT) for safety-critical autonomous UAS systems.

However, an adversarial code and semantic traceability audit of `scripts/verify_downstream_baseline.py` (lines 750-760) and `skills/spec-orchestrator/` reveals a critical quality gate omission across upstream and downstream MBSE pipelines:

### 1. Absence of Ground-Truth Provenance Gate in Conformance Verifier (lines 750-760)
In `scripts/verify_downstream_baseline.py`, the `_run_verification()` function orchestrates baseline quality gates:
```python
# Run Checks 10, 11, 12, 13, 14, 15, 16, and 17
check_gitignore_exists(repo_root)
check_no_ds_store_files(repo_root)
check_no_duplicate_master_blueprints(dest)
check_latex_katex_syntax(repo_root)
check_downstream_instructions_exist(repo_root)
check_reconcile_backlog_tooling_exists(repo_root)
check_upstream_template_clean_landing_zones(repo_root)
check_safety_integrity_and_sora_completeness(repo_root)
```
While Checks 10 through 17 rigorously enforce file existence, LaTeX math syntax, clean landing zones, and SORA/STPA table schemas, the baseline verifier completely lacks a **Check 18: Ground-Truth Provenance & Anti-Hallucination Gate**. Consequently, specifications with severe aerodynamic and parametric contradictions pass baseline verification with exit code 0.

### 2. Failure Modes Permitted by the Absent Gate
Without automated ground-truth provenance enforcement, autonomous specification subagents introduce three categories of correctness-breaking defects:

1. **Platform Parametric Drift**: Subagents hallucinate contradictory physical and aerodynamic attributes across markdown specifications and SysML models:
   - Wingspan: Hallucinating 2.2 m or 2.4 m instead of the certified 1.80 m airframe invariant.
   - MTOW: Hallucinating $< 25$ kg or 22.0 kg instead of the certified 17.0 kg platform MTOW.
   - Stall Airspeed ($V_s$): Stating 18.0 m/s or 20.0 m/s instead of the aerodynamic stall limit of 24.0 m/s.
   - Cruise ($V_{cruise}$) and Max ($V_{max}$) Speeds: Drifting to 25.0 m/s and 50.0 m/s instead of SSOT values 31.0 m/s and 42.0 m/s.
   - Catapult Launch Pressure: Permitting out-of-range pressures (e.g. 5.0 bar or 20.0 bar) outside the certified [8.5, 14.0] bar envelope.

2. **Ungrounded AI / Deep-Learning Buzzword Infiltration**: Subagents inject ungrounded machine-learning assumptions (e.g., CUDA GPU kernel acceleration, PyTorch models, YOLOv8 vision transformers, runtime gradient backpropagation, deep reinforcement learning flight control) into flight-critical guidance, navigation, and control specifications. In safety-critical DO-178C Level B avionics (e.g., STM32H7 / Cortex-M7 dual-core microcontrollers), GPU runtimes and non-deterministic neural networks cannot be certified or hosted. Real-world optical tracking relies on deterministic classical algorithms (2D Kalman filter centroid tracking with Mahalanobis innovation gating).

3. **Unresolved Citation Placeholders & Broken Schema Links**: Subagents leave raw template placeholders (such as `[link-to-schema]` or `[link-to-specification]`) or broken relative links in markdown `## Source References` blocks, severing bidirectional traceability between functional requirements and normative ground-truth engineering documents.

### 3. Missing Parity Auditor Validator in `skills/spec-orchestrator/`
The upstream specification orchestrator in `skills/spec-orchestrator/parity_auditor/src/parity_auditor/validators/` contains structural and syntax linters (`link_validator.py`, `mermaid_syntax_validator.py`, `katex_validator.py`), but lacks an integrated `ground_truth_validator.py`. As a result, Phase 1 through Phase 4 specification generation workflows lack early-stage hallucination detection.

**Invariants Violated**:
- **Single Source of Truth (SSOT) Mandate** (`.pipeline/constitution.md` Section Downstream Single Source of Truth Mandate): Downstream specifications and derived SysML models must preserve 100% parametric parity with authoritative platform schemas (`schema/Avenger5.sysml`).
- **Pure Schema-Driven Compiler Invariant** (`.pipeline/constitution.md` Section Core System Boundaries & Invariants): Agents are strictly forbidden from inventing, proposing, or hardcoding ungrounded domain concepts into pipeline specifications.
- **Traceability Integrity Mandate** (`.pipeline/constitution.md` Section Traceability): Every Feature and User Story must maintain valid, resolvable citations to authoritative ground-truth sources.
- **Safety Criticality & Certifiability Invariant** (DO-178C Level B, ASTM F3269-17, SORA SAIL II): Control law and tracking architectures must specify deterministic, certifiable algorithms rather than unverified deep learning frameworks.

## 4. UML Diagrams

```mermaid
sequenceDiagram
    participant LLM as SpecSubagent
    participant ORC as SpecOrchestrator
    participant VER as BaselineVerifier
    participant HIL as EmbeddedHILTestbench
    participant GTV as GroundTruthScanner

    LLM->>ORC: Submits specification with hallucinated CUDA and 2.4m wingspan
    ORC->>VER: Runs baseline conformance checks 10 through 17
    Note over VER: Checks 10-17 pass syntax and markdown formatting
    
    alt Baseline Without Ground-Truth Scanner
        VER-->>ORC: Conformance gate passes with Exit Code 0 (False Positive)
        ORC->>HIL: Synthesizes embedded C++ control law from ungrounded spec
        Note over HIL: Hardware watchdog trip - CUDA unsupported on STM32H7 MCU
        HIL-->>ORC: Embedded build crash and DO-178C certification rejection
    else Core Pipeline With Ground-Truth Provenance Scanner
        VER->>GTV: Executes Check 18 Ground-Truth Provenance Audit
        Note over GTV: Scans parametric invariants and ungrounded AI buzzwords
        GTV-->>VER: Flags wingspan 2.4m and CUDA buzzword (Exit Code 1)
        VER-->>ORC: Conformance gate halts deployment before code synthesis
    end
```

## 5. Affected Callers / Downstream Impact

- **Downstream Application Repositories (`uas-003`, `gintatkinson/uas-*`)** — Downstream engineering projects risk compiling ungrounded specifications that contradict physical airframe dimensions, leading to invalid flight dynamics models, incorrect aerodynamic coefficients, and flawed mass properties in simulation.
- **Embedded Flight Control Law Synthesis (MATLAB / Simulink / Stateflow / Embedded Coder)** — Control law autogeneration produces destabilizing control gains when parameterized with hallucinated aerodynamic stall speeds ($V_s = 18.0$ m/s instead of $24.0$ m/s) or incorrect vehicle mass properties ($MTOW = 25.0$ kg instead of $17.0$ kg).
- **Run-Time Assurance (RTA) Safety Net Monitors (ASTM F3269-17, DO-178C Level B)** — Safety monitor boundary trip conditions depend on precise aerodynamic envelope parameters. Parametric inaccuracies in specification baselines directly compromise RTA failsafe switching thresholds.
- **Upstream Core Distribution Engine (`DEAP-spec-core`, `scripts/install_pipeline.sh`)** — Upstream installer templates omit `scripts/audit_hallucinations.py` from the core tooling distribution package, forcing downstream teams to manually create ad-hoc verification scripts.

## 6. Proposed Correction

Integrate `scripts/audit_hallucinations.py` into the upstream Core MBSE pipeline, package it in `scripts/install_pipeline.sh`, and bind it as **Check 18** in `scripts/verify_downstream_baseline.py` and `parity_auditor`.

### 6.1 Check 18 Implementation in `scripts/verify_downstream_baseline.py`

```python
def check_ground_truth_provenance(repo_root: str) -> None:
    """Check 18: Ground-Truth Provenance & Anti-Hallucination Quality Gate.
    
    Audits downstream specifications and SysML v2 models against platform invariants:
    1. Parametric consistency (wingspan=1.8m, MTOW=17.0kg, Vs=24.0m/s, Vcruise=31.0m/s, Vmax=42.0m/s).
    2. Zero ungrounded AI / Deep Learning buzzwords (CUDA, YOLO, PyTorch, TensorRT, Backprop).
    3. Authoritative ground-truth citations to schema/extracted/ and schema/Avenger5.sysml.
    """
    upstream_marker = os.path.join(repo_root, ".pipeline", "upstream")
    if os.path.isdir(upstream_marker):
        print("Success: Check 18 verified (Upstream template detected — skipping concrete ground-truth scan).")
        return

    scanner_script = os.path.join(repo_root, "scripts", "audit_hallucinations.py")
    if not os.path.isfile(scanner_script):
        print(f"ERROR: Check 18 failed: Ground-truth scanner '{scanner_script}' missing.", file=sys.stderr)
        sys.exit(1)

    cmd = [sys.executable, scanner_script, "--repo-root", repo_root]
    res = subprocess.run(cmd, capture_output=True, text=True)
    if res.returncode != 0:
        print(f"ERROR: Check 18 failed (Ground-Truth Provenance & Anti-Hallucination violations):\n{res.stdout}", file=sys.stderr)
        sys.exit(1)

    print("Success: Check 18 verified (Ground-Truth Provenance & Anti-Hallucination Scanner passed with 0 errors).")
```

### 6.2 Binding Check 18 in `_run_verification()`

```python
def _run_verification(args, dest, repo_root, is_flutter, is_react):
    # Run Checks 10, 11, 12, 13, 14, 15, 16, 17, and 18
    check_gitignore_exists(repo_root)
    check_no_ds_store_files(repo_root)
    check_no_duplicate_master_blueprints(dest)
    check_latex_katex_syntax(repo_root)
    check_downstream_instructions_exist(repo_root)
    check_reconcile_backlog_tooling_exists(repo_root)
    check_upstream_template_clean_landing_zones(repo_root)
    check_safety_integrity_and_sora_completeness(repo_root)
    check_ground_truth_provenance(repo_root)
```

### 6.3 Packaging in `scripts/install_pipeline.sh`

```bash
# Package ground-truth tooling and test suites into downstream workspaces
cp -RP "$INSTALLER_ROOT/scripts/audit_hallucinations.py" "$TARGET_DIR/scripts/" 2>/dev/null || true
cp -RP "$INSTALLER_ROOT/tests/test_audit_hallucinations.py" "$TARGET_DIR/tests/" 2>/dev/null || true
```

## 7. Relationship to Existing Issues

Discovered in audit — new finding.

## Audit Source

Adversarial Semantic Traceability Audit
SEVERITY: Critical
FILE_LOCATION: scripts/verify_downstream_baseline.py:750-760

**Comment by gintatkinson** [comment]
Resolved and verified in upstream compiler core.
